File: java/src/main/java/org/github/salpadding/tinywasm/InstanceImpl.java
package org.github.salpadding.tinywasm;

public class InstanceImpl implements Instance{
    private final int descriptor;

    public InstanceImpl(int descriptor) {
        this.descriptor = descriptor;
    }

    @Override
    public byte[] getMemory(int off, int len) {
//        return Natives.getMemory(this.descriptor, off, len);
        return null;
    }

    @Override
    public void setMemory(int off, byte[] buf) {
//        Natives.setMemory(this.descriptor, off, buf);
    }

    @Override
    public long[] execute(String export, long[] args) {
        return Natives.execute(this.descriptor, export, args);
    }

    @Override
    public void close() throws Exception {
//        Natives.close(this.descriptor);
    }
}

File: java/src/main/java/org/github/salpadding/wasmer/InstanceImpl.kt
package org.github.salpadding.wasmer

import kotlin.concurrent.withLock


internal class InstanceImpl(private val descriptor: Int) : Instance {
    override val id: Int
        get() = descriptor

    override fun getMemory(off: Int, len: Int): ByteArray {
        if (off < 0 || len < 0) {
            throw RuntimeException("off or len shouldn't be negative")
        }
        return Natives.getMemory(this.descriptor, off, len)
    }

    override fun setMemory(off: Int, buf: ByteArray) {
        if (off < 0) {
            throw RuntimeException("off shouldn't be negative")
        }
        Natives.setMemory(this.descriptor, off, buf)
    }

    override fun execute(export: String, args: LongArray): LongArray {
        return Natives.execute(descriptor, export, args)
    }

    override fun close() {
        Natives.GLOBAL_LOCK.withLock {
            Natives.HOSTS[descriptor] = null
            Natives.close(descriptor)
        }
    }
}